scalyr_agent/copying_manager/copying_manager.py

```diff
@@ -1340,7 +1341,33 @@ def __consolidate_checkpoints(self, checkpoints):
             if os.path.isfile(active_checkpoint_path):
                 os.unlink(active_checkpoint_path)
 
-    def __find_and_read_checkpoints(self, warn_on_stale=False):
+    def _get_all_checkpoints(self):  # type: () -> Dict[six.text_type, Any]
+        """
+        Collects all checkpoint states from all running workers and the checkpoints from saved consolidated file.
+        :return: dict of the checkpoints.
+        """
+        checkpoints = {}  # type: Dict[six.text_type, Any]
+
+        # also get the previously consolidated file.
+        consolidated_checkpoint_path = os.path.join(
+            self.__config.agent_data_path, CONSOLIDATED_CHECKPOINTS_FILE_NAME
+        )
+
+        # also read existing checkpoints from the consolidated checkpoint file.
+        previous_checkpoints = self.__read_checkpoint_state(
+            consolidated_checkpoint_path
+        )
+        if previous_checkpoints:
+            checkpoints.update(previous_checkpoints["checkpoints"])
+
+        # get all checkpoints from all files that are currently processed by all workers.
+        for worker in self.workers.values():
+            for worker_session in worker.sessions:
+                checkpoints.update(worker_session.get_full_checkpoints())
```

Comment:
What happens if there are multiple checkpoints for a given file?  (For example, the consolidated checkpoint file has one checkpoint, but the worker has a newer one.)  It looks like from the sequence, we will always take whatever the worker gives us as the latest version.  However, I wonder if we should explicitly make sure that the timestamp in that checkpoint is newer than what was in the consolidated one.

Would doing that check be useful?  I guess because we always consolidate before we start the workers, there is no real way the worker checkpoint could be older than the consolidated one.  Is that true?

Reply:
Yes, I think that's true. The new checkpoint object is always created from the state of the current log processor. Even if the consolidated checkpoint file has already had the saved state for the same log file, this state should be used as the initial position for the log processor when it has been created, so the new checkpoint from the worker should contain at least the same position(in case if a log file hasn't been written) as the consolidated log file.